vm/devices/storage/storvsc_driver/src/lib.rs

```diff
@@ -0,0 +1,913 @@
+                            Ok(()) => Ok(()),
+                            Err(err) => {
+                                tracing::error!(
+                                    "Unable to send new request to VMBus, err={:?}",
+                                    err
+                                );
+                                Err(err)
+                            }
+                        }
+                    }
+                    Err(err) => {
+                        tracing::error!("Unable to receive new request, err={:?}", err);
+                        Err(StorvscError(StorvscErrorInner::RequestError))
+                    }
+                },
+                Event::VmbusPacketReceived(result) => match result {
+                    Ok(packet_ref) => self.handle_packet(packet_ref.as_ref()),
+                    Err(err) => {
+                        tracing::error!("Error receiving VMBus packet, err={:?}", err);
+                        Err(StorvscError(StorvscErrorInner::Queue(err)))
+                    }
+                },
+            }?;
+        }
+    }
+
+    fn send_request<M: RingMem>(
+        &mut self,
+        request: &storvsp_protocol::ScsiRequest,
+        buf_gpa: u64,
+        byte_len: usize,
+        writer: &mut queue::WriteHalf<'_, M>,
+        completion_sender: Sender<StorvscCompletion>,
+    ) -> Result<(), StorvscError> {
+        // Create pending transaction record
+        let transaction_id = self
+            .transactions
+            .insert(PendingOperation::new(completion_sender));
+
+        self.send_gpa_direct_packet(
+            writer,
+            storvsp_protocol::Operation::EXECUTE_SRB,
+            storvsp_protocol::NtStatus::SUCCESS,
+            transaction_id as u64,
+            request,
+            buf_gpa,
+            byte_len,
+        )
+    }
+
+    async fn cancel_pending_completions(&mut self) {
+        for transaction in self.transactions.iter_mut() {
+            transaction.1.cancel();
+        }
+        self.transactions.clear();
+    }
+
+    fn handle_packet<M: RingMem>(
+        &mut self,
+        packet: &IncomingPacket<'_, M>,
+    ) -> Result<(), StorvscError> {
+        let packet = parse_packet(packet)?;
+        match packet {
+            Packet::Data(data) => {
+                match data.operation {
+                    storvsp_protocol::Operation::ENUMERATE_BUS => {
+                        // Nothing to do here, and no completion is required
+                        Ok(())
+                    }
+                    _ => Err(StorvscError(StorvscErrorInner::UnexpectedOperation)),
+                }
+            }
+            Packet::Completion(completion) => {
+                // Parse ScsiRequest (contains response) from bytes
+                let result =
+                    storvsp_protocol::ScsiRequest::ref_from_bytes(completion.data.as_slice())
+                        .map_err(|_err| StorvscError(StorvscErrorInner::UnexpectedOperation))?
```

Comment:
nit: in this case, the error message will be a bit unhelpful. Have you considered something more specific? Presumably you are getting an error here because `completion.data` does not contain sufficient length to be converted to a `ScsiRequest`, implying that the packet is too small (or the wrong packet)?